iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -882,17 +873,24 @@ private async Task EnsureOpenedAsync(bool withRetry, CancellationToken cancellat
 
         private async Task EnsureOpenedAsync(bool withRetry, TimeoutHelper timeoutHelper)
         {
-            if (Volatile.Read(ref _opened))
+            using var cts = new CancellationTokenSource(timeoutHelper.GetRemainingTime());
+            using var operationCts = CancellationTokenSource.CreateLinkedTokenSource(cts.Token, _cancelPendingOperationsCts.Token);
+
+            // If this object has already been disposed, we will throw an exception indicating that.
+            // This is the entry point for interacting with the client and this safety check should be done here.
+            // The current behavior does not support open->close->open
+            if (_isDisposed)
             {
-                return;
+                throw new ObjectDisposedException(nameof(RetryDelegatingHandler));
             }
 
-            bool gain = await _clientOpenCloseSemaphore.WaitAsync(timeoutHelper.GetRemainingTime()).ConfigureAwait(false);
-            if (!gain)
+            if (Volatile.Read(ref _opened))
```

Comment:
Consider what we read about Eric Lippert's take on `volatile` and instead use `Interlocked.Exchange`.